apps/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts

```diff
@@ -373,19 +374,44 @@ export class WorkspaceInstallManager extends BaseInstallManager {
   protected async postInstallAsync(): Promise<void> {
     // Grab the temp shrinkwrap, as this was the most recently completed install. It may also be
     // more up-to-date than the checked-in shrinkwrap since filtered installs are not written back.
-    const tempShrinkwrapFile: BaseShrinkwrapFile = ShrinkwrapFileFactory.getShrinkwrapFile(
+    // Note that if there are no projects, or if we're in PNPM workspace mode and there are no
+    // projects with dependencies, a lockfile won't be generated.
+    const tempShrinkwrapFile: BaseShrinkwrapFile | undefined = ShrinkwrapFileFactory.getShrinkwrapFile(
       this.rushConfiguration.packageManager,
       this.rushConfiguration.pnpmOptions,
       this.rushConfiguration.tempShrinkwrapFilename
-    )!;
-
-    // Write or delete all project shrinkwraps related to the install
-    await Promise.all(
-      this.rushConfiguration.projects.map(async (project) => {
-        await tempShrinkwrapFile.getProjectShrinkwrap(project)?.updateProjectShrinkwrapAsync();
-      })
     );
 
+    if (tempShrinkwrapFile) {
+      // Write or delete all project shrinkwraps related to the install
+      await Async.forEachAsync(
+        this.rushConfiguration.projects,
+        async (project) => {
+          await tempShrinkwrapFile.getProjectShrinkwrap(project)?.updateProjectShrinkwrapAsync();
+        },
+        { concurrency: 30 }
+      );
+    } else if (
+      this.rushConfiguration.packageManager === 'pnpm' &&
+      this.rushConfiguration.pnpmOptions?.useWorkspaces
+    ) {
+      // If we're in PNPM workspace mode and PNPM didn't create a shrinkwrap file,
+      // there are no dependencies. Generate empty shrinkwrap files for all projects.
+      await Async.forEachAsync(
+        this.rushConfiguration.projects,
+        async (project) => {
+          await BaseProjectShrinkwrapFile.saveEmptyProjectShrinkwrapFileAsync(project);
+        },
+        { concurrency: 30 }
```

Comment:
`{ concurrency: 30 }` seems a bit arbitrary.  Shouldn't we have a global parallelism that defaults to the number of CPU cores?

Reply:
Discussed offline. Changed it to 10.